GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java

```diff
@@ -0,0 +1,349 @@
+                executeOnce = false;
+            }
+        }
+        else
+        {
+            emit(particlePositions, particleVelocities, particleTimeStamps);
+        }
+
+    }
+
+    /**
+     * Append the bounding volume particle positions, times and velocities to the existing mesh
+     * before creating a new scene object with this mesh attached to it.
+     * Also, append every created scene object and its creation time to corresponding array lists.
+     *
+     * @param particlePositions
+     * @param particleVelocities
+     * @param particleTimeStamps
+     */
+
+    private void emit(float[] particlePositions, float[] particleVelocities,
+                      float[] particleTimeStamps)
+    {
+        float[] allParticlePositions = new float[particlePositions.length + particleBoundingVolume.length];
+        System.arraycopy(particlePositions, 0, allParticlePositions, 0, particlePositions.length);
+        System.arraycopy(particleBoundingVolume, 0, allParticlePositions,
+                        particlePositions.length, particleBoundingVolume.length);
+
+
+        float[] allSpawnTimes = new float[particleTimeStamps.length + BVSpawnTimes.length];
+        System.arraycopy(particleTimeStamps, 0, allSpawnTimes, 0, particleTimeStamps.length);
+        System.arraycopy(BVSpawnTimes, 0, allSpawnTimes, particleTimeStamps.length, BVSpawnTimes.length);
+
+        float[] allParticleVelocities = new float[particleVelocities.length + BVVelocities.length];
+        System.arraycopy(particleVelocities, 0, allParticleVelocities, 0, particleVelocities.length);
+        System.arraycopy(BVVelocities, 0, allParticleVelocities, particleVelocities.length, BVVelocities.length);
+
+
+        Particles particleMesh = new Particles(mGVRContext, mMaxAge,
+                mParticleSize, mEnvironmentAcceleration, mParticleSizeRate, mFadeWithAge,
+                mParticleTexture, mColor);
+
+
+        GVRSceneObject particleObject = particleMesh.makeParticleMesh(allParticlePositions,
+                allParticleVelocities, allSpawnTimes);
+
+        this.addChildObject(particleObject);
+        mActiveParticleMeshes.add(particleObject);
+        mParticleMeshSpawnTimes.add(currTime);
+    }
+
+
+    /**
+     * Create a bouding volume for the particle system centered at its position with
+     * the specified width, height and depth. This is important to do because the parent scene
+     * object might fall outside the viewing frustum and cause the entire system to be
+     * culled.
+     * This function creates 8 particles (mesh vertices) with very large spawning time
+     * attributes (i.e. they are always discarded) which define the volume of the system. The
+     * system is assumed to stay inside this volume.
+     * @param width volume length (along x-axis)
+     * @param height volume height (along y-axis)
+     * @param depth volume depth (along z-axis)
+     */
+
+    public void setParticleVolume(float width, float height, float depth)
```

Comment:
Since this can be called from any method it is unlikely to be thread-safe. In this particular case you could just run it on the gl thread.